dohproxy/utils.py

```diff
@@ -346,3 +353,24 @@ def configure_logger(name='', level='DEBUG'):
         raise Exception("Invalid log level name : %s" % level_name)
     logger.setLevel(level)
     return logger
+
+
+def set_dns_ecs(dnsq, ip):
+    """Sets RFC 7871 EDNS Client Subnet (ECS) options in a DNS packet.
+    :param dnsq: DNS packet.
+    :param ip: IP address. String or ipaddress object.
+    """
+    if not isinstance(
+        ip,
+        (ipaddress.IPv4Address, ipaddress.IPv6Address)
+    ):
+        ip = ipaddress.ip_address(ip)
+    ip_supernet_bits = 56 if ip.version == 6 else 24
+    ip_supernet = ipaddress.ip_network(ip).supernet(
+        new_prefix=ip_supernet_bits,
+    )
+    dnsq.use_edns()
+    dnsq.options.append(dns.edns.ECSOption(
+        address=ip_supernet.network_address.compressed,
+        srclen=ip_supernet_bits,
+    ))
```

Comment:
My understanding is that this will clear the original EDNS payload. I think it should be done in a way similar to `want_dnssec`: https://github.com/rthalley/dnspython/blob/e31b10472822afd8e22906fb6b837cbc2ca3b962/dns/message.py#L534-L548

e.g
1. check if edns RR is already there. if not, set it with `use_edns()`
1. set ECS if it was not previously set (a client can request to not send ECS by setting it to 0.0.0.0/0) (which is what I mentioned in #27)
1. remove ECS on the way out (unless it was set by client originally)
1. remove EDNS if it was not set by client originally.

Reply:
Sorry, didn't even notice issue #27! I actually whipped this one up after showing doh-proxy to a friend (who works for a major CDN and hence has an interest in ECS being everywhere) who pointed me in the direction of ECS.

Huh, I thought EDNS options stacked (i.e. were not limited to one per option), but apparently not.

The chain you specified makes sense; I'll get that implemented.  use_edns() is idempotent, but as we need to know the old value anyway for (potentially) changing in back in the reply, it can be part of the check.

Reply:
> Huh, I thought EDNS options stacked (i.e. were not limited to one per option), but apparently not.

well, you can have multiple EDNS options within the EDNS Resource Record.
https://tools.ietf.org/html/rfc6891#section-6.1.1

> The OPT RR MAY be placed anywhere within the additional data section.
   When an OPT RR is included within any DNS message, it MUST be the
   only OPT RR in that message.